### PR TITLE
handle plugin manifests automatically

### PR DESCRIPTION
**Pull request by dkanada**
I made this small to hopefully backport it and avoid a manifest requirement in a stable release. It's a bit of a hack but the manifests can easily be generated from the repository information (including the new thumbnail images) to avoid duplicate metadata.

### CODE REVIEW

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 1]
```diff
@@ -332,34 +334,54 @@ public void FailPlugin(Assembly assembly)
...
         }
 
+        /// <inheritdoc/>
+        public bool GenerateManifest(PackageInfo packageInfo, Version version, string path)
+        {
+            var versionInfo = packageInfo.Versions.First(v => v.Version == version.ToString());
+            var url = packageInfo.ImageUrl ?? string.Empty;
+            var imageFilename = url.Substring(url.LastIndexOf('/') + 1, url.Length);
+
+            using (var client = new WebClient())
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:360]: We use `HttpClient`

Something like:
```c#
            await using var fileStream = File.OpenWrite(Path.Join(path, imageFilename));
            var downloadStream = await _httpClientFactory.CreateClient(NamedClient.Default)
                .GetStreamAsync(url)
                .ConfigureAwait(false);
            await downloadStream.CopyToAsync(fileStream).ConfigureAwait(false);
```

You can get `IHttpClientFactory` in the `PluginManager` constructor

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 2]
```diff
@@ -332,34 +334,54 @@ public void FailPlugin(Assembly assembly)
...
+            {
+                Category = packageInfo.Category,
+                Changelog = versionInfo.Changelog ?? string.Empty,
+                Description = packageInfo.Description,
+                Id = new Guid(packageInfo.Id),
+                Name = packageInfo.Name,
+                Overview = packageInfo.Overview,
+                Owner = packageInfo.Owner,
+                TargetAbi = versionInfo.TargetAbi ?? string.Empty,
+                Timestamp = DateTime.Parse(versionInfo.Timestamp ?? string.Empty),
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:375]:
```suggestion
                Timestamp = string.IsNullOrEmpty(versionInfo.Timestamp) ? DateTime.MinValue : DateTime.Parse(versionInfo.Timestamp)
```

`DateTime.Parse` can't handle `string.Empty`

File: Emby.Server.Implementations/Plugins/PluginManager.cs (hunk 1 above)
Comment (dkanada) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:360]: This requires async which the PluginManager doesn't use yet, do you want it added?

Comment (crobibero) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:360]: If it's not too much work, the function should be `async` as well.
If too many other changes are required `.GetAwaiter().GetResult()` can be used with a `TODO` note

Comment (dkanada) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:360]: This whole pull request is a TODO until we add these fields to the database :laughing: but I'll add the async now anyways. It's marked as a draft because I haven't tested it as thoroughly as possible yet, any chance you have any edge cases you want me to confirm?

Comment (crobibero) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:360]: I would test valid config, config without image, and config where the image url isn’t an image (folder or other random file)

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 3]
```diff
@@ -332,34 +351,68 @@ public void FailPlugin(Assembly assembly)
...
+            var versionInfo = packageInfo.Versions.First(v => v.Version == version.ToString());
+            var imagePath = string.Empty;
+
+            if (!string.IsNullOrEmpty(packageInfo.ImageUrl))
+            {
+                var url = new Uri(packageInfo.ImageUrl);
+                imagePath = Path.Join(path, url.Segments.Last());
+
+                await using var fileStream = File.OpenWrite(imagePath);
+                var downloadStream = await HttpClientFactory
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:387]:
```suggestion
                await using var downloadStream = await HttpClientFactory
```

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 4]
```diff
@@ -332,34 +351,68 @@ public void FailPlugin(Assembly assembly)
...
+                    .CreateClient(NamedClient.Default)
+                    .GetStreamAsync(url)
+                    .ConfigureAwait(false);
+
+                await downloadStream.CopyToAsync(fileStream).ConfigureAwait(false);
+                await fileStream.DisposeAsync();
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:393]:
```suggestion
```

This is already being disposed, no need to manually call
(I know Rider suggests adding this, but it's not needed)

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 5]
```diff
@@ -332,34 +351,68 @@ public void FailPlugin(Assembly assembly)
...
+
+            if (!string.IsNullOrEmpty(packageInfo.ImageUrl))
+            {
+                var url = new Uri(packageInfo.ImageUrl);
+                imagePath = Path.Join(path, url.Segments.Last());
+
+                await using var fileStream = File.OpenWrite(imagePath);
+                var downloadStream = await HttpClientFactory
+                    .CreateClient(NamedClient.Default)
+                    .GetStreamAsync(url)
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:389]: Should catch the `HttpRequestException` so plugin processing doesn't stop if the image is invalid
When catching the exception, I would also clear `imagePath` so it's not set to an invalid location

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 6]
```diff
@@ -332,34 +351,68 @@ public void FailPlugin(Assembly assembly)
...
+                return false;
+            }
+
+            var versionInfo = packageInfo.Versions.First(v => v.Version == version.ToString());
+            var imagePath = string.Empty;
+
+            if (!string.IsNullOrEmpty(packageInfo.ImageUrl))
+            {
+                var url = new Uri(packageInfo.ImageUrl);
+                imagePath = Path.Join(path, url.Segments.Last());
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:384]:
```suggestion
                imagePath = Path.Join(path, url.Segments[^1]);
```

**Review by crobibero — APPROVED** [review]

**Review by BaronGreenback — APPROVED** [review]

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 7]
```diff
@@ -332,32 +351,74 @@ public void FailPlugin(Assembly assembly)
...
         {
-            if (manifest == null)
-            {
-                return false;
-            }
-
             try
             {
                 var data = JsonSerializer.Serialize(manifest, _jsonOptions);
                 File.WriteAllText(Path.Combine(path, "meta.json"), data);
```
Comment (Bond-009) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:360]: Can't we make `SaveManifest` async so you can serialize to a stream?
If not, `SerializeToUtf8Bytes` should be faster

Comment (dkanada) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:360]: I tried to avoid refactoring since this will be backported, that seems like something that can go in a new pull request.

**Review by Bond-009 — APPROVED** [review]
